DescendPeaks in C++

Pascallio opened the pull request:
This is the C++ implementation of the descendPeaks function of MSnbase. It takes a slightly different approach to the original by first subsetting the maximum region to consider. Next for each left/right descend, it checks if the next signal is A) above the signal percentage threshold and B) not higher than the previous signal (rising). If the signal passes both tests, the left/right index is updated. After both descends, these indexes are used to define the region of which the weighted mean of the mass is taken.

It is recommended to sanatize all inputs first in R before calling this function to ensure the right values are used, e.g. C-indexes, not R-indexes.

Pascallio committed:
DescendPeaks in C++

This is the C++ implementation of the descendPeaks function of MSnbase. It takes a slightly different approach to the original by first subsetting the maximum region to consider. Next for each left/right descend, it checks if the next signal is A) above the signal percentage threshold and B) not higher than the previous signal (rising). If the signal passes both tests, the left/right index is updated. After both descends, these indexes are used to define the region of which the weighted mean of the mass is taken.

lgatto commented:
Just a quick comment - it would also be important to have unit tests to check that the results give expected results on small data or/and same as an R implementation on a more complex data.

sgibb commented on src/descendPeak.cpp:
```diff
@@ -0,0 +1,143 @@
+using namespace Rcpp;
+using namespace std;
+
+double weightedMean(NumericVector mzs, NumericVector ints) {
+    // Get the total intensity
+    double intensity = sum(ints);
+
+    // Calculate the weighted mass
+    int n = ints.length();
+    double total = 0;
+    for (int i = 0; i < n; i++) {
+        total += mzs[i] * ints[i] / intensity;
+    }
+    return total;
+}
```
Without testing I would assume that you could reduce the runtime of this function by avoid looping over the vector twice (sum has to go through all indices as well):
```suggestion
double weightedMean(NumericVector mzs, NumericVector ints) {
    int n = ints.length();
    double total = 0;
    double intensity = 0;
    for (int i = 0; i < n; i++) {
        total += mzs[i] * ints[i];
        intensity += ints[i];
    }
    return total / intensity;
}
```

Pascallio replied:
I agree, this improves the weighted mean function.

sgibb commented:
@Pascallio Thanks for your contribution. It is very welcome! Maybe a stupid question/suggestion. We try to keep the dependencies as low as possible (Suggests contains already a large list of dependencies). Currently we successfully avoid `Rcpp` as direct dependency (some packages in Suggests of course already depend on `Rcpp`). It would be great if you ensure that `Rcpp` would end in `Suggests` and not in `Depends` or `Imports` or (implement this whole stuff in pure C).

sgibb commented on src/descendPeak.cpp:
```diff
@@ -0,0 +1,143 @@
+               NumericVector masses, double signalRatio) {
+
+    // Maximum deviation from the centroid index
+    int maxK = 30;
```
Depending on the solution of the spectra an window of +/- 30 may be to large/small. I don't like to hardcode window sizes. I would suggest to make it an argument as well.

Pascallio replied:
Agreed, I will push an update that adds the `maxK` parameter to the function arguments.

sgibb commented on src/descendPeak.cpp:
```diff
@@ -0,0 +1,143 @@
+        bool goodSignal =  current / centroidValue > signalRatio;
+
+        // Compare current to previous value to check for a rising signal
+        bool risingSignal = current >= previous;
```
Would this be a good test? Already a small plateau (2 indices of the same intensity value) would break here, even if it goes down after the overnext index.

Pascallio replied:
I do believe that is the default behaviour in the original MSnbase version here: https://github.com/lgatto/MSnbase/blob/1338323345d2a6cf856f1819dfb365300ab6d9d5/R/functions-Spectrum.R#L781-L788

I'm not sure about the current version in MsCoreUtils, but I could implement the `stopAtTwo` behaviour or perhaps stop at any given `k` if desired?

Pascallio committed:
Remove the need for sum

Avoids looping over the intensity vector twice

Co-authored-by: Sebastian Gibb <mail@sebastiangibb.de>

Pascallio committed:
Moved maxK to function arguments

Allows for setting a different maxK, which determines the maximum region the descendPeak algorithm is applied to

Pascallio commented:
> @Pascallio Thanks for your contribution. It is very welcome! Maybe a stupid question/suggestion. We try to keep the dependencies as low as possible (Suggests contains already a large list of dependencies). Currently we successfully avoid `Rcpp` as direct dependency (some packages in Suggests of course already depend on `Rcpp`). It would be great if you ensure that `Rcpp` would end in `Suggests` and not in `Depends` or `Imports` or (implement this whole stuff in pure C).

Thank you @sgibb!  I fully understand the concern about dependencies. I don't have much experience with Rcpp, so how do I ensure it would end up in `Suggests`? Does that mean I should replace the Rcpp `Range` and change NumericVectors with for instance std::vector? 

As for implementing it in pure C, I have no experience with C so it might take me a while to submit a pure C implementation but I'm willing to give it a shot when I have some spare time.

lgatto commented:
Not sure if this PR is still relevant, but I would like to see a unit test, that compares the R- and C-level results.
